src/java.base/share/classes/sun/net/www/protocol/jar/JarURLConnection.java

```diff
@@ -88,8 +89,14 @@ public void close () throws IOException {
             try {
                 super.close();
             } finally {
-                if (!getUseCaches()) {
-                    jarFile.close();
+                try {
+                    if (!getUseCaches()) {
+                        jarFile.close();
+                    }
+                } finally {
+                    if (jarFileURLConnection instanceof FileURLConnection fileURLConnection) {
```

Comment:
See discussion at https://github.com/openjdk/jdk/pull/12871#discussion_r1198932229.

Reply:
> should maybe the metadata function initializeHeaders close its own stream

Another possibility is for `FileURLConnection.getHeaderField(String)` and related methods to never open a `FileInputStream` to begin with, amending https://github.com/openjdk/jdk/blob/8474e693b4404ba62927fe0e43e68b904d66fbde/src/java.base/share/classes/sun/net/www/protocol/file/FileURLConnection.java#L96-L100 That would require overriding at least https://github.com/openjdk/jdk/blob/8474e693b4404ba62927fe0e43e68b904d66fbde/src/java.base/share/classes/sun/net/www/URLConnection.java#L104-L131 since https://github.com/openjdk/jdk/blob/8474e693b4404ba62927fe0e43e68b904d66fbde/src/java.base/share/classes/sun/net/www/protocol/file/FileURLConnection.java#L137-L140 etc. call `super`. The legality of such a change in light of https://github.com/openjdk/jdk/blob/8474e693b4404ba62927fe0e43e68b904d66fbde/src/java.base/share/classes/java/net/URLConnection.java#L57-L66 https://github.com/openjdk/jdk/blob/8474e693b4404ba62927fe0e43e68b904d66fbde/src/java.base/share/classes/java/net/URLConnection.java#L92-L96 is unclear; the Javadoc implies but does not state that you _cannot_ access header fields before `connect()` is called. On the other hand https://github.com/openjdk/jdk/blob/8474e693b4404ba62927fe0e43e68b904d66fbde/src/java.base/share/classes/sun/net/www/protocol/jar/JarURLConnection.java#L232-L235 apparently presumes that any request for a header field implicitly `connect()`s if it needed to.

Reply:
That seems like it could cause other issues down the road though. `FileURLConnection::connect` has this comment:

```
    /*
     * Note: the semantics of FileURLConnection object is that the
     * results of the various URLConnection calls, such as
     * getContentType, getInputStream or getContentLength reflect
     * whatever was true when connect was called.
     */
    public void connect() throws IOException {
```

Reply:
This patch appears to be work and may be preferable to what is currently proposed:

```diff
diff --git src/java.base/share/classes/sun/net/www/protocol/file/FileURLConnection.java src/java.base/share/classes/sun/net/www/protocol/file/FileURLConnection.java
index a9ce7cee9a1..55783474c22 100644
--- src/java.base/share/classes/sun/net/www/protocol/file/FileURLConnection.java
+++ src/java.base/share/classes/sun/net/www/protocol/file/FileURLConnection.java
@@ -53,7 +53,6 @@ public class FileURLConnection extends URLConnection {
     File file;
     String filename;
     boolean isDirectory = false;
-    boolean exists = false;
     List<String> files;
 
     long length = -1;
@@ -93,16 +92,12 @@ public class FileURLConnection extends URLConnection {
     private boolean initializedHeaders = false;
 
     private void initializeHeaders() {
-        try {
-            connect();
-            exists = file.exists();
-        } catch (IOException e) {
-        }
-        if (!initializedHeaders || !exists) {
+        if (!initializedHeaders || !file.exists()) {
             length = file.length();
             lastModified = file.lastModified();
+            filename = file.toString();
 
-            if (!isDirectory) {
+            if (!file.isDirectory()) {
                 FileNameMap map = java.net.URLConnection.getFileNameMap();
                 contentType = map.getContentTypeFor(filename);
                 if (contentType != null) {
@@ -131,17 +126,17 @@ public class FileURLConnection extends URLConnection {
 
     public Map<String,List<String>> getHeaderFields() {
         initializeHeaders();
-        return super.getHeaderFields();
+        return properties.getHeaders();
     }
 
     public String getHeaderField(String name) {
         initializeHeaders();
-        return super.getHeaderField(name);
+        return properties.findValue(name);
     }
 
     public String getHeaderField(int n) {
         initializeHeaders();
-        return super.getHeaderField(n);
+        return properties.getKey(n);
     }
 
     public int getContentLength() {
```

Reply:
Ah! It's true only if the first time connect is called implicitly by `initializeHeaders()` (sigh)...

Reply:
Hello Jesse, Daniel, I had a look at the existing code in `sun.net.www.protocol.file.FileURLConnection` its super class `sun.net.www.URLConnection` and then its superclass `java.net.URLConnection`.

The various header related methods in `sun.net.www.URLConnection` first call the `getInputStream()` (and ignore that stream completely) and then check the `properties` field to get the header value. The call to `getInputStream()` is an internal implementation detail (since the `java.net.URLConnection` public APIs make no mention that it's required or will be called) and it seems to me that it's being called to make sure that the underlying resource is "connectable/available". The returned `InputStream` itself never gets used.

For the `FileURLConnection` class, the `InputStream` plays no role to populate these header fields. Yet, before the changes in this PR, it does end up creating an `InputStream` when trying to populate these header fields. I think the cleanest fix would be to not create this `InputStream` when populating these header fields (i.e. from `initializeHeaders()` method). That would mean that the `initializeHeaders()` wouldn't need to call the `connect()` method, like what's being proposed in this discussion thread. I think not calling `connect()` from `initializeHeaders()` (which gets called from various getHeader... methods) will still satisfy the specification because `java.net.URLConnection#connect()` method states:

>
>...
> Operations that depend on being connected, like getContentLength, will implicitly perform the connection, if necessary.

The "if necessary" in that sentence I think allows us to not "connect()" when populating and returning header fields.

I think the only place we should call connect() internally in this `FileURLConnection` is when `getInputStream()` method gets called on this class.

P.S: The current implementation in `initializeHeaders()` has some oddities. For example, if the `File` was a directory then in `connect()` we set `isDirectory()` as `true`. Then at some later point, when someone calls a getHeader... method and `initializeHeaders()` gets called, if the directory was deleted from the filesystem, then `initializeHeaders()` `exists` will be false, so it will go into the `if (!initializedHeaders || !exists) {` block but will still end up using the old outdated `isDirectory` flag to populate the `properties`. That's a different unrelated issue though.

Reply:
I believe we should wait for @Michael-Mc-Mahon input before reworking these methods to not call `connect()`. There seems to be an assumption that `connect()` would/should be called.
What I like with the current fix is that it's a very narrow fix which is unlikely to cause too much backward compatibility issues.